### PR TITLE
Fix #1951 Optimise hints_summary.xml

### LINKED ISSUE #1951: Optimise hints_summary.xml
This issue is tracking the optimisation work of `hints_summary.xml`

1. Make sure that the layout uses `ConstraintLayout` and there is minimal nesting of views. If this view already uses `ConstraintLayout` then it's fine.

2. Make sure that all textviews use the style from `styles.xml` file like `Heading1`, `TextFieldLabel`, `Body`, etc. whichever suits best.

While making the above changes the UI should not get affected and the end result should still be same as before (or an optimised version of UI).

If there are multiple files with the same name, update them all.

## References
Mobile Portrait: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/grid/
Mobile Landscape: https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/grid/
Tablet: https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/grid/
Wiki: https://github.com/oppia/oppia-android/wiki/Working-on-UI

### LINKED ISSUE #3537: Merge hints_summary.xml into single xml file
Currently there are 4 versions of `hints_summary.xml` file merge it into single xml file.

We can use https://text-compare.com/ to compare two versions of this file and for all the differences we can create variables in `dimens.xml` file and use it accordingly.

It can happen that it becomes really hard to merge all 4 files into single file, in such cases prefer code-readability over single file. Therefore it can be fine if in the end the code still contains 2 or 3 files.

**Note**: **In PR, make sure you add before and after screenshot of mobile-portrait, mobile-landscape, tablet-portrait and tablet-landscape for comparison and make sure that there is not difference between before and after UI.**

### PR DESCRIPTION
**Pull request by ArchitJain1201**
## Explanation

Fix #1951
Fix #3537 Merge hints_summary.xml into single xml file 
## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

**Before**

**After**
3 1a
**Before**

**After**
3 2a
**Before**

**After**
1 2a
**Before**

**After**
2 1a
**Before**

**After**
2 2a
**Before**

**After**
1 1a

### CODE REVIEW

**Comment by ArchitJain1201** [comment]
## Changes Made
- Optimize hints_summary.xml
- Merged hints_summary.xml files into Single xml file 
- changed the textSize from 18sp to 16sp in style for Heading4

**Comment by ArchitJain1201** [comment]
Problem
- App is crashing in Tablet mode not able to find the Reason
- Not able to use contraintLayout in `@+id/hint_summary_container` and `@+id/hint_list_container `as `@+id/hint_list_container` require to be constraint which is disturbing the UI
- Not able to add `app:layout_constraintStart_toEndOf="@+id/hint_title"` in `FrameLayout @+id/expand_list_icon` as Tablet mode don't require it.

PTAl @rt4914  @bkaur-bkj

File: app/src/main/res/layout/hints_summary.xml [hunk 1]
```diff
@@ -94,21 +89,17 @@
...
-        android:layout_height="wrap_content"
-        android:fontFamily="sans-serif"
-        android:textColor="@color/oppia_primary_text"
-        android:textSize="16sp" />
+        style="@style/Heading4"
+        android:layout_gravity="start" />
     </LinearLayout>
   </LinearLayout>
-</layout>
+</layout>
```
Comment (bkaur-bkj) [inline app/src/main/res/layout/hints_summary.xml]: @ArchitJain1201 add a newline here

File: app/src/main/res/values-sw600dp-land/dimens.xml [hunk 2]
```diff
@@ -420,4 +420,24 @@
...
+  <dimen name="hint_summary_reveal_hint_button_layout_marginBottom">28dp</dimen>
+  <dimen name="hint_summary_expand_list_icon_layout_marginStart">0dp</dimen>
+  <dimen name="hint_summary_expand_list_icon_layout_marginEnd">0dp</dimen>
+  <dimen name="hint_summary_expand_list_icon_layout_marginTop">32dp</dimen>
+  <dimen name="hint_summary_expand_list_icon_layout_marginBottom">16dp</dimen>
+  <dimen name="hint_summary_expand_list_icon_layout_constraintHorizontal_bias">0.5</dimen>
+  <dimen name="hint_summary_expand_list_icon_layout_constraintVertical_bias">0.5</dimen>
+  <dimen name="hint_summary_hint_list_container_layout_marginStart">176dp</dimen>
+  <dimen name="hint_summary_hint_list_container_layout_marginEnd">176dp</dimen>
+  <dimen name="hint_summary_hint_list_container_layout_marginTop">0dp</dimen>
```
Comment (bkaur-bkj) [inline app/src/main/res/values-sw600dp-land/dimens.xml]: @ArchitJain1201 newline above `</resources>` like it was before, also do it in all the dimens file too.

**Review by bkaur-bkj — CHANGES_REQUESTED** [review]
@ArchitJain1201 PTAL at these minor changes, also regarding the problem mentioned I am also not sure why it crashes, @rt4914 PTAL

File: app/src/main/res/values/styles.xml [hunk 3]
```diff
@@ -247,7 +247,7 @@
 
   <style name="Heading4" parent="TextView.Common1">
     <item name="android:fontFamily">sans-serif</item>
-    <item name="android:textSize">18sp</item>
+    <item name="android:textSize">16sp</item>
```
Comment (bkaur-bkj) [inline app/src/main/res/values/styles.xml:254]: @ArchitJain1201 also I think this heading 4 value should not be altered. Can you pls explain the reason for this change?

Comment (ArchitJain1201) [reply app/src/main/res/values/styles.xml:254]: as Heading1 have textsize =24, Heading2 have textsize= 20, Heading3 have textsize= 18sp and Heading5 have textsize=14sp. So, I do think the textSize of Heading4 should be 16sp

File: app/src/main/res/layout/hints_summary.xml (hunk 1 above)
Comment (ArchitJain1201) [reply app/src/main/res/layout/hints_summary.xml]: done

File: app/src/main/res/values-sw600dp-land/dimens.xml (hunk 2 above)
Comment (ArchitJain1201) [reply app/src/main/res/values-sw600dp-land/dimens.xml]: done

**Comment by ArchitJain1201** [comment]
PTAL @rt4914 @bkaur-bkj

**Comment by rt4914** [comment]
Will review it after @bkaur-bkj 's approval.

**Comment by bkaur-bkj** [comment]
@rt4914 I just have doubt regarding the change of heading size rest it LGTM

**Comment by ArchitJain1201** [comment]
PTAL @rt4914

**Review by rt4914 — CHANGES_REQUESTED** [review]
@ArchitJain1201 & @bkaur-bkj Apologies for delay. 

This PR has been done with great efforts but unfortunately I will have to close this PR and issue.

Reason: As I see the changes in this PR it looks that combining these xml files is really complicated, especially because of usage of `bias` in `mobile-portrait` and not in any other file. As a result I think that the better solution is to actually keep these files separate because even a minor change in any of the design for this screen will end up create a lot of confusion. But if we keep 4 files it will not be confusing it will just require extra efforts.

What are you thoughts?

**Comment by ArchitJain1201** [comment]
@rt4914 Sir, I do not think `bias` is a problem as by default bias is 0.5 (as far as I know) and I have assigned 0.5 in tablet mode but even if it is a problem then we can make it into two .XML file one for mobile and other for the tablet mode.(as basis is not used in tablet mode)

**Comment by ArchitJain1201** [comment]
@rt4914 please also look at the changes I made for `Heading4` in `style` file. And I have also given the reasoning for it.

**Comment by ArchitJain1201** [comment]
PTAL @rt4914

**Comment by bkaur-bkj** [comment]
> @ArchitJain1201 & @bkaur-bkj Apologies for delay.
> 
> This PR has been done with great efforts but unfortunately I will have to close this PR and issue.
> 
> Reason: As I see the changes in this PR it looks that combining these xml files is really complicated, especially because of usage of `bias` in `mobile-portrait` and not in any other file. As a result I think that the better solution is to actually keep these files separate because even a minor change in any of the design for this screen will end up create a lot of confusion. But if we keep 4 files it will not be confusing it will just require extra efforts.
> 
> What are you thoughts?

Yes as you said confusion can be created because of design changes, what are your reviews about the point mentioned by @ArchitJain1201 on bias?

**Comment by rt4914** [comment]
@ArchitJain1201 Your point is valid. If you can merge this in two files one for mobile (portrait + landscape) and second for tablet (portrait + landscape). I think that would be great. But do keep this point is mind: while making changes if you think that the changes are complicated and might create confusion in future then I suggest not to work on that further and instead close the PR/issue.

**Comment by ArchitJain1201** [comment]
PTAL @rt4914 @bkaur-bkj

**Comment by ArchitJain1201** [comment]
@rt4914 Sir, I have made two XML files one for mobile and the other for tablet versions. But it is still crashing for tablet version can you please check it. Please also look at the `Heading4` in the `style` file. I have provided the reasoning above on the @bkaur-bkj comment.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@ArchitJain1201 Couple of points:
1. Have you tried these [steps](https://github.com/oppia/oppia-android/wiki/Frequent-Errors-and-Solutions#crashing-layout-tags-in-tablet) to fix the error. If the error still occurs can you mention the error.
2. Your Thought process regarding Heading4 is correct but we cannot make this change because Heading4 is getting used at other places too and it will change the UI for other screens. I suggest undo this change.

3. @bkaur-bkj Can you help Archit arrange the screenshot side-by-side for comparison in before/after screenshots. In current format they are hard to compare.

**Pull request closed without merging**